Sd ia estimate next purchase

sandaiiyahh opened the pull request:
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

- Added new variables to `itemTemplate`:  `numberofPurchases` and `daysUntilPurchase`
- `numberOfPurchases` is the number of times an item has been purchased. Any time an item is checked, this value increments by 1.
- `daysUntilPurchase` is the calculated result of the `calculateEstimate` function
- Declared a `latestInterval` initially set to `frequency`, which will be used inside the `calculateEstimate` function. `latestInterval` is the current date minus the previously purchased date, in the form of # of days.
- Used `calculateEstimate` inside the `handleChange` function 
- Stored milliseconds conversion into a variable `mlsPerDay`
- Changed firestore from `.set` to `.update` to save having to set variables each time

Thoughts We Had: 
- Should `frequency` be used when calculating the estimate?
- Not sure if the `daysUntilPurchase` updates accurately. Will need to calculate on our own to test it

UPDATE: 
- Used `frequency` as initial `latestInterval` value to be used in `calculateEstimate` function
- Added a condition inside the `handleChange` to check if the checkbox is checked before using the `calculateEstimate` function so that it doesn't update when the checkbox is both check _and_ unchecked
- `daysUntilPurchase` is now set to `nextPurchaseDate` _or_ `daysUntilPurchase` -- `daysUntilPurchase` so that it doesn't update when the box is unchecked

<!-- What does this code change? Why did I choose this approach? Did I learn anything worth sharing? Reminder: This will be a publicly facing representation of your work (READ: help you land that sweet dev gig). -->

## Related Issue

closes #10 

<!-- If you write "closes" followed by the Github issue number, it will automatically close the issue for you when the PR merges -->

## Acceptance Criteria

- [x] When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database
- [x] The script at `/src/lib/estimates.js` should be used to calculate the next estimated purchase interval.

<!-- Include AC from the Github issue -->

## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓  | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before

<img width="371" alt="Screen Shot 2021-08-10 at 9 01 33 PM" src="https://user-images.githubusercontent.com/60532744/128953939-aae87215-b3c0-4737-aff4-ccbaa9679577.png">

<!-- If UI feature, take provide screenshots -->

### After

Variables are initially declared when you first add an item: 
<img width="352" alt="Screen Shot 2021-08-10 at 9 02 14 PM" src="https://user-images.githubusercontent.com/60532744/128953988-3b1a06ce-7727-4f6d-abdf-855519bce54b.png">

After checking off an item as purchased:
<img width="404" alt="Screen Shot 2021-08-10 at 9 03 07 PM" src="https://user-images.githubusercontent.com/60532744/128954028-f766c077-717e-45d0-a1af-ec6ff1aa0ec0.png">

After an item has been purchased more than 2 times:
<img width="397" alt="Screen Shot 2021-08-10 at 9 04 22 PM" src="https://user-images.githubusercontent.com/60532744/128954121-5c42345b-e0ab-4ca4-968c-13b367e4bb96.png">

<!-- If UI feature, take provide screenshots -->

## Testing Steps / QA Criteria

- Pull down branch `sd-ia-estimate-next-purchase`
- Add a new item
- Check off item to mark as purchased
- Change `lastPurchasedDate` of item in firestore to a date before the previous purchase
- Once item has reached over 2 purchases, the `daysUntilPurchase` should change

<!-- Provide steps the other cohort members and mentors need to follow to properly test your additions. -->

isaabutaa committed:
add numberOfPurchases and previousEstimate as values to the itemTemplate

Commit:
add daysUntilPurchase to itemTemplate in AddItem

Commit:
pull  in values from AddItem to SingleItem as props

isaabutaa committed:
WIP: Trying to calculate next estimate for purchase, numberOfPurchases NAN

Commit:
remove previousEstimate variable in AddItem

Commit:
change updateIsPurchased firestore set to update

Commit:
remove comments

Commit:
use variable mlsPerDay in SingleItem

Commit:
change setInterval time back to minute

ksiman14 commented on src/components/SingleItem.js:
```diff
@@ -1,45 +1,62 @@
-        token,
-        name,
-        frequency,
+        numberOfPurchases: numberOfPurchases + 1,
```
This is causing a small bug - The number of purchases updates both when you check and uncheck an item. We may want to add a condition that numberOfPurchases only increments when isPurchased (pre-update) is false (i.e. when isPurchased is about to turn to true).

isaabutaa replied:
Oh, good catch, @ksiman14 ! I wonder if we could do something like this:

`numberOfPurchases: !isPurchased ? numberOfPurchases + 1 : numberOfPurchases`

@sandaiiyahh, do you think that would work?

sandaiiyahh replied:
Ooh, yes thanks for catching that Kelsey! Yeah, let's try that and let us know if that works @isaabutaa !

ksiman14 commented on src/components/SingleItem.js:
```diff
@@ -1,45 +1,62 @@
     }
   });
 
+  let latestInterval = 0;
```
Based on the equation in estimates.js, should latestInterval be initialized as Number(frequency)? And then leave the following condition as is, so we only update it from the frequency if there is a last purchase date. That way if we have not purchased an item two or more times, there still is a non-zero estimated nextPurchaseDate (which is the frequency).

isaabutaa replied:
@ksiman14 thanks for bringing this up. We were wondering where `frequency` would fit in as well. @segdeha clarified in Slack that `frequency` _is_ one of the inputs for the `calculateEstimate` function. I think it makes sense that we could input that value as the argument for `latestInterval`. I'll test it locally and see how it changes things. @sandaiiyahh 👀

sandaiiyahh replied:
Hmm, if we're going by its definition, latestInterval is the number of days between the most recent and the previous purchase. The frequency is more of a user input of how often they would want to buy an item, so I'm not sure 😅. Yeah, let's see how the numbers are like when you make the changes, @isaabutaa.

ksiman14 reviewed (changes requested):
@isaabutaa and @sandaiiyahh as we all talked about during office hours, this was a _really_ tough issue, so good job! 

After testing, I have encountered a bug and had one additional idea. I'm leaving individual comments on some lines with my thoughts.

hulyak reviewed (commented):
Good job Sandy and Isa,  I tried to check if the `numberOfPuchases` would update after changing the `lastPurchaseDate` and uncheck the item. It is working and counting the number of purchases every time.

isaabutaa committed:
bring in frequency to SingleItem component through props, only increment number of purchases if item was previously not purchased, initialize latestInterval value to frequency

isaabutaa committed:
Only set daysUntilPurchase as the nextPurchaseDate if the checkbox is checked. If it's not, leave it as the previous value.

isaabutaa committed:
Put calculateEstimate function inside condition statement so that it calculates the estimate only if the checkbox is checked. Set daysUntilPurchase value to nextPurchaseDate if checked, or daysUntilPurchased if not checked

isaabutaa committed:
Merge pull request #33 from the-collab-lab/experimenting-with-estimate

Some small changes based on comments in sd-ia-estimate-next-purchase PR 👍

luisaugusto commented on src/components/SingleItem.js:
```diff
@@ -1,45 +1,68 @@
-        name,
-        frequency,
+        numberOfPurchases: !isPurchased
+          ? numberOfPurchases + 1
```
Since older items don't have this field, it would be good to include a fallback:

```suggestion
          ? numberOfPurchases + 1 || 1
```

luisaugusto reviewed (commented):
Hey, you all did an amazing job getting this set up! I'm glad you were all able to work with the `calculateEstimate` function, I know that can be confusing, but your calculations were great, and everything is being entered into Firestore correctly. Nice work!

There is one issue where if you uncheck an item because you accidentally clicked, there is no way to undo the calculation that you already made, so you can keep toggling the checkbox to alter the data for the item with no way to revert. 

I noticed that there are currently three fields in Firebase: `isPurchased`, `lastPurchasedDate`, and `numberOfPurchases`. What I would recommend doing is consolidating those three fields into one `purchaseDates` array. That way, you can:

- Add a new date to the array when an item is checked, and remove it if it's unchecked (recalculate the `daysUntilPurchase` when toggling.
- Calculate if the item is purchased based on if the most recent purchase date is further than the `daysUntilPurchase` calculation
- Count the number of purchases based on the length of the array

If you have any questions about this, we can discuss it together on Slack 😀 I'll be available tomorrow and Saturday and will try to respond as quickly as I can so that we can have this merged in by Saturday!

---

Other than that, there are just a couple of small items:
- For older items in lists that were made before this change, the `numberOfPurchases` field is returning `NaN`, it would be good to check this and default it to `0` if it's missing in an older item without the field.
- I just recently noticed that `frequency is being saved as a `string`, which sets the `daysUntilPurchase` field to a string after running `calculateEstimate` for the first time. I don't think this is causing any bugs, but there might be some type errors later on in the application if we decide to manipulate this more.

meganesu commented:
Nicely done y'all! Like others have said, this was a complicated ticket, and you did a great job sticking with it! Well done 🥳 

A few thoughts in response to @luisaugusto's review: (happy to discuss more tomorrow in Slack!)

> There is one issue where if you uncheck an item because you accidentally clicked, there is no way to undo the calculation that you already made, so you can keep toggling the checkbox to alter the data for the item with no way to revert.

Personally, in the interest of time (since I know y'all have already spent quite a while working on this), I'd suggest we create another story to handle undoing accidental purchase calculations, since that functionality is not explicitly included in the acceptance criteria for this ticket.

> I noticed that there are currently three fields in Firebase: `isPurchased`, `lastPurchasedDate`, and `numberOfPurchases`. What I would recommend doing is consolidating those three fields into one `purchaseDates` array.

Similarly, I got the impression from @segdeha's [comment in Slack](https://the-collab-lab.slack.com/archives/C023GB5D72N/p1628640192046600?thread_ts=1628639520.046300&cid=C023GB5D72N) that storing a list of all the purchase dates is also out of scope.

> For older items in lists that were made before this change, the `numberOfPurchases` field is returning `NaN`, it would be good to check this and default it to `0` if it's missing in an older item without the field.

Agreed, adding a quick check and default value is a good idea. During office hours, we'd debated dumping the existing documents in the database and starting from scratch (since it's a small project and we don't have any real users yet). But in a real professional setting, that wouldn't be an option, since odds are you'd already have a bunch of user data in the database.

luisaugusto reviewed (approved):
@meganesu okay no problem! We can create a separate issue for it 🚀

isaabutaa committed:
Update src/components/SingleItem.js

Update line 63 of SingleItem component so that numberOfPurchases equals 1 for items in the database that did not previously have a numberOfPurchases value and are purchased for the first time

Co-authored-by: Luis Augusto <hello@luis.app>

isaabutaa committed:
Ensure frequency is of type Number()

isaabutaa commented:
Hey @meganesu @luisaugusto thank-you both for the feedback and suggestions. Since it was rather small, I committed the change that Luis suggested:
- changed ` ? numberOfPurchases + 1 : numberOfPurchases` to ` ? numberOfPurchases + 1 || 1 : numberOfPurchases`
- Also implemented `Number(frequency)` in a couple places to ensure that `frequency` is saved as a Number going forward
